includes/WcfceMailProcessor.inc

```diff
@@ -0,0 +1,223 @@
+   */
+  public function createActivity() {
+    if (!$this->isMessageValid()) {
+      return FALSE;
+    }
+
+    if (!civicrm_initialize()) {
+      return FALSE;
+    }
+
+    $assignee_id = $this->getContactIdByEmail($this->message['to']);
+    $creator_id  = $this->getContactIdByEmail($this->message['from']);
+
+    if (!$creator_id) {
+      return FALSE;
+    }
+
+    $eml = $this->composeEml();
+    $cases = &$this->message['params']['submission']->civicrm['case'];
+
+    // For all cases related to the webform:
+    foreach ($cases as $case) {
+      // Create activity.
+      $result = civicrm_api3('Activity', 'create', [
+        'sequential' => 1,
+        'activity_type_id' => 'Email',
+        'case_id' => $case['id'],
+        'subject' => $this->message['subject'],
+        'details' => trim($this->message['body'][0]),
+        'assignee_id' => $assignee_id,
+        'source_contact_id' => $creator_id,
+      ]);
+      // Attach eml file to it.
+      if (!empty($result['values'][0]['id'])) {
+        $this->attachEmlToActivity($result['values'][0]['id'], $eml);
+      }
+    }
+
+    return $result['values'][0] ?? [];
+  }
+
+  /**
+   * Attach eml file to activity.
+   *
+   * @param int $activity_id
+   *   Activity id.
+   * @param string $eml
+   *   Contents of eml file.
+   * @param array|NULL $message
+   *   Email data. If is set to NULL then $this->message would be used.
+   *
+   * @return array|false
+   *   Attached file entity data, empty array or FALSE in case of error.
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function attachEmlToActivity($activity_id, $eml, $message = NULL) {
+    if (!$activity_id) {
+      return FALSE;
+    }
+
+    $result = civicrm_api3('Attachment', 'create', [
+      'sequential' => 1,
+      'mime_type' => 'message/rfc822',
+      'entity_table' => 'civicrm_activity',
+      'entity_id' => $activity_id,
+      'content' => $eml,
+      'name' => $this->generateEmlFileName($message),
+    ]);
+
+    // Set file type to attached file.
+    // (The file_type_id db table column is always NULL and not used in civicrm
+    // at all. But we need to set some flag to differentiate original email from
+    // other attached files (and disable edit/delete for this file later), so we
+    // will use that field for this purpose and just set 1 as a value)
+    if (!empty($result['values'][0]['id'])) {
+      civicrm_api3('File', 'create', [
+        'id' => $result['values'][0]['id'],
+        'file_type_id' => 1,
+      ]);
+    }
+
+    return $result['values'][0] ?? [];
+  }
+
+  /**
+   * Generate name for eml file.
+   *
+   * @param array|NULL $message
+   *   Email data. If is set to NULL then $this->message would be used.
+   *
+   * @return string
+   */
+  public function generateEmlFileName($message = NULL) {
+    $message = $message ?? $this->message;
+    $res = date('Ymd_Hi_')
+      . preg_replace('!\W+!', '', trim(strtolower($message['subject'])))
```

Comment:
What if the message subject gets really long. Maybe we should cut it off at some particular character length?